### ISSUE TITLE
Flow 模型finetune 会逐渐崩掉?

### ISSUE THREAD

**Issue by dyyoungg** [FunAudioLLM/CosyVoice#633]
我按照 processor的流程写了一份online 数据准备的训练流程，核心就是利用 S3tokenizer的pytorch版进行在线speech token推理（https://github.com/xingchensong/S3Tokenizer 感谢这个repo，验证后推理结果没问题）。Flow 训练需要5个输入，分别是speech token， speech_token_len, speech_feat, speech_feat_len，还有spk_embedding,  除了speech token，其他的都放到 dataset去做，代码
如下。
 ```
import onnxruntime
from functools import partial
from torch.utils.data import Dataset
import torchaudio.compliance.kaldi as kaldi
import numpy as np
import torch
import torchaudio
import librosa
import torch.nn.functional as F

class VocosDataset(Dataset):
    def __init__(self, file):
        self.filelist = []
       
        self.filelist = parse_filelist(file)
       
        self.sampling_rate = 16000
        self.num_samples = 64000
        self.win_length = 1024

        option = onnxruntime.SessionOptions()
        option.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_ALL
        option.intra_op_num_threads = 1
        self.campplus_session = onnxruntime.InferenceSession("./cosyvoice/campplus.onnx", 
                                                             sess_options=option, 
                                                             providers=["CPUExecutionProvider"])
    
        self.feat_extractor=partial(mel_spectrogram, # matcha tts
                n_fft=1024,
                num_mels=80,
                sampling_rate=22050,
                hop_size=256,
                win_size=1024,
                fmin=0,
                fmax=8000,
                center=False,
                )
    
    def __len__(self) -> int:
        return len(self.filelist)
    
    def resample_audio(self, speech, sr):

        prompt_speech_22050 = torchaudio.transforms.Resample(orig_freq=sr, new_freq=22050)(speech)
        max_val = torch.max(torch.abs(prompt_speech_22050))
        if max_val > 1.0:
            prompt_speech_22050 = prompt_speech_22050 / max_val
        return prompt_speech_22050

    def _extract_spk_embedding(self, speech):

        feat = kaldi.fbank(speech,
                           num_mel_bins=80,
                           dither=0,
                           sample_frequency=16000)
        feat = feat - feat.mean(dim=0, keepdim=True)
        embedding = self.campplus_session.run(None,
                                              {self.campplus_session.get_inputs()[0].name: feat.unsqueeze(dim=0).cpu().numpy()})[0].flatten().tolist()
        embedding = torch.tensor([embedding])
        return embedding
    
    def __getitem__(self, index: int) -> torch.Tensor:
        audio_path = self.filelist[index]
        start = 0
        sample = {}
        y1, sr = librosa.load(audio_path, sr=self.sampling_rate, mono=True)
        y = torch.tensor(y1).float().unsqueeze(0)
        if y.size(0) > 1:
            # mix to mono
            y = y.mean(dim=0, keepdim=True)

         # extract embedding
        embedding = self._extract_spk_embedding(y)
        # normalize
        normalized_embedding = F.normalize(embedding.squeeze(), dim=0) # 不normal 第一步后效果就有问题
        sample["spk_embedding"] = normalized_embedding
  
        if y.size(-1) < self.num_samples:
            sample["speech_token_len"] = torch.tensor(int(y.size(-1) // 320), dtype=torch.long)# 50hz
            prompt_speech_22050 = self.resample_audio(y, sr)
            mel_feature = self.feat_extractor(prompt_speech_22050).squeeze(dim=0) # [80, T]
            
            sample['speech_feat_len'] = mel_feature.size(-1)

            # pad
            y = torch.concat([y, torch.zeros(1, self.num_samples - y.shape[1])], dim=1)
            target_frames = int(22050 / 16000 * self.num_samples // 256)  # 344
            mel_feature = torch.nn.functional.pad(mel_feature, (0, target_frames - mel_feature.size(-1)))
            sample["speech_feat"] = mel_feature
            
        else:
            start = np.random.randint(low=0, high=y.size(-1) - self.num_samples + 1)
            y = y[:, start : start + self.num_samples]
            sample["speech_token_len"] = torch.tensor(int(y.size(-1) // 320), dtype=torch.long)# 50hz
            prompt_speech_22050 = self.resample_audio(y, sr)
            
            mel_feature = self.feat_extractor(prompt_speech_22050).squeeze(dim=0) # [80, T]
            sample['speech_feat_len'] = mel_feature.size(-1)
            sample["speech_feat"] = mel_feature

        sample["speech_16k"] = y.squeeze()

        # {"speech", "speech_token_len", "speech_feat", "speech_feat_len"， “spk_embedding”}

        return sample
```

然后speech token 在训练时单独去推，推完根据speech_token_len 去mask 0

```
for batch in dataloader:
     speech_token = token_extractor(audio_input_16k) # [bs, T]
     max_len = speech_token.size(1)
     speech_token_len = batch["speech_token_len"].cuda(local_rank)
     mask = torch.arange(max_len, device=speech_token.device).expand(speech_token.size(0), max_len) < speech_token_len.unsqueeze(1)
     speech_token = (speech_token * mask).to(dtype=torch.long)
```
感觉整个流程没啥问题，但全量finetune会逐渐变差，大概200步之后，验证集重建的mel谱就变为噪声类型，前100步结果都还可以接受，但也是逐渐变差, 验证集损失逐渐变大，训练的loss 倒是一直比较平稳的下降，比较奇怪。模型部分是保持官方的代码训练流程，并且加载了 cosyvoice sft版本的flow模型。
不知道是不是我遗漏了什么数据处理的细节，还是说 是因为全量微调就不太行？数据质量应该是没问题，因为在其他模型上验证过。微调设置 lr=2e-5, warmup=2500, bs=32。希望大佬能解答一下。

**Comment by aluminumbox** [comment]
是否验证过Pytorch模型推理结果与onnx一致？这个pytorch模型应该freeze住不要参与训练

**Comment by dyyoungg** [comment]
> 是否验证过Pytorch模型推理结果与onnx一致？这个pytorch模型应该freeze住不要参与训练

验证过，推理结果是一致的，tokenizer是freeze住的。
```
class S3TokenExtract(nn.Module):
    def __init__(self, ) -> None:
        super().__init__()
        self.speech_tokenizer = s3tokenizer.load_model("speech_tokenizer_v1.onnx")
        self.speech_tokenizer.freeze()

    @torch.no_grad()
    def forward(self, speech):
        feat = whisper.log_mel_spectrogram(speech, n_mels=128).to(speech.device)
        feat_len = torch.tensor([feat.shape[-1]]*feat.shape[0], dtype=torch.int32).to(speech.device)
        speech_codes, speech_codes_lens = self.speech_tokenizer.quantize(feat, feat_len) 
        speech_codes = speech_codes.clone()  # for backward compatbility
        speech_codes_lens = speech_codes_lens.clone()  # for backward compatbility
        return speech_codes
```

**Comment by dyyoungg** [comment]
找到问题了，是我在build 模型时 有问题，，原repo 加上我这套在线 token 推理的方式 是没问题的，感谢。

> 是否验证过Pytorch模型推理结果与onnx一致？这个pytorch模型应该freeze住不要参与训练

**Comment by dukGuo** [comment]
> 找到问题了，是我在build 模型时 有问题，，原repo 加上我这套在线 token 推理的方式 是没问题的，感谢。
> 
> > 是否验证过Pytorch模型推理结果与onnx一致？这个pytorch模型应该freeze住不要参与训练

方便告知一下是哪方面的问题吗？我似乎也碰到同样的情况，感谢🙏

**Comment by JohnHerry** [comment]
finetune 一般几个epoch？
